WcaOnRails/app/webpacker/components/EditSchedule/store/reducer.js

```diff
@@ -0,0 +1,216 @@
+          ...room,
+          activities: [
+            ...room.activities,
+            {
+              ...payload.wcifActivity,
+              id: nextActivityId(state.wcifSchedule),
+            },
+          ],
+        }) : room)),
+      })),
+    },
+  }),
+
+  [EditActivity]: (state, { payload }) => ({
+    ...state,
+    wcifSchedule: {
+      ...state.wcifSchedule,
+      venues: state.wcifSchedule.venues.map((venue) => ({
+        ...venue,
+        rooms: venue.rooms.map((room) => ({
+          ...room,
+          activities: room.activities.map((activity) => (
+            activity.id === payload.activityId
+              ? { ...activity, [payload.key]: payload.value }
+              : activity
+          )),
+        })),
+      })),
+    },
+  }),
+
+  [RemoveActivity]: (state, { payload }) => ({
+    ...state,
+    wcifSchedule: {
+      ...state.wcifSchedule,
+      venues: state.wcifSchedule.venues.map((venue) => ({
+        ...venue,
+        rooms: venue.rooms.map((room) => ({
+          ...room,
+          activities: room.activities.filter((activity) => activity.id !== payload.activityId),
+        })),
+      })),
+    },
+  }),
+
+  [MoveActivity]: (state, { payload }) => ({
+    ...state,
+    wcifSchedule: {
+      ...state.wcifSchedule,
+      venues: state.wcifSchedule.venues.map((venue) => ({
+        ...venue,
+        rooms: venue.rooms.map((room) => ({
+          ...room,
+          activities: room.activities.map((activity) => (
+            activity.id === payload.activityId
+              ? moveActivityByDuration(activity, payload.isoDuration)
+              : activity
+          )),
+        })),
+      })),
+    },
+  }),
+
+  [ScaleActivity]: (state, { payload }) => ({
+    ...state,
+    wcifSchedule: {
+      ...state.wcifSchedule,
+      venues: state.wcifSchedule.venues.map((venue) => ({
+        ...venue,
+        rooms: venue.rooms.map((room) => ({
+          ...room,
+          activities: room.activities.map((activity) => (
+            activity.id === payload.activityId
+              ? scaleActivitiesByDuration(activity, payload.isoDeltaStart, payload.isoDeltaEnd)
+              : activity
+          )),
+        })),
+      })),
+    },
+  }),
+
+  [EditVenue]: (state, { payload }) => ({
+    ...state,
+    wcifSchedule: {
+      ...state.wcifSchedule,
+      venues: state.wcifSchedule.venues.map((venue) => (venue.id === payload.venueId ? {
+        ...venue,
+        [payload.propertyKey]: payload.newProperty,
+        rooms: venue.rooms.map((room) => ({
+          ...room,
+          activities: (
+            payload.propertyKey === 'timezone'
+              ? room.activities.map((activity) => (
+                changeActivityTimezone(
+                  activity,
+                  venue.timezone,
+                  payload.newProperty,
+                )
+              ))
+              : room.activities
+          ),
+        })),
```

Comment:
It would be ideal if objects/arrays which don't get any actual updates could keep the same reference, rather than creating a new identical object/array. That way if they're ever used in a dependency array it won't unexpectedly trigger the callback/effect.
In this case that would mean moving `payload.propertyKey === 'timezone'` up outside of the rooms map.
I don't think any other reducers here have such an issue (unless we let users try and delete ids which don't exist).

Reply:
Ah, now I got it. Thanks for your explanation, fixed!